beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1715,13 +1739,37 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
 
         // Store the block and its state, and execute the confirmation batch for the intermediate
         // states, which will delete their temporary flags.
+        // If the write fails, revert fork choice to the version from disk, else we can
+        // end up with blocks in fork choice that are missing from disk.
+        // See https://github.com/sigp/lighthouse/issues/2028
         ops.push(StoreOp::PutBlock(
             block_root,
             Box::new(signed_block.clone()),
         ));
         ops.push(StoreOp::PutState(block.state_root, &state));
         let txn_lock = self.store.hot_db.begin_rw_transaction();
-        self.store.do_atomically(ops)?;
+
+        if let Err(e) = self.store.do_atomically(ops) {
+            error!(
+                self.log,
+                "Database write failed!";
+                "msg" => "Restoring fork choice from disk",
+                "error" => ?e,
+            );
+            match Self::load_fork_choice(self.store.clone())? {
+                Some(persisted_fork_choice) => {
+                    *fork_choice = persisted_fork_choice;
+                }
+                None => {
+                    crit!(
+                        self.log,
+                        "No stored fork choice found to restore from";
+                        "warning" => "The database is likely corrupt now"
```

Comment:
Perhaps we should mention `--purge-db` here?

Reply:
Done!